// @nrwl/workspace/src/command-line/workspace-integrity-checks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const output_1 = require("./output");
const shared_1 = require("./shared");
class WorkspaceIntegrityChecks {
    constructor(projectNodes, files) {
        this.projectNodes = projectNodes;
        this.files = files;
    }
    run() {
        return [...this.projectWithoutFilesCheck(), ...this.filesWithoutProjects()];
    }
    projectWithoutFilesCheck() {
        const errors = this.projectNodes
            .filter(n => n.files.length === 0)
            .map(p => `Cannot find project '${p.name}' in '${p.root}'`);
        const errorGroupBodyLines = errors.map(f => `${output_1.output.colors.gray('-')} ${f}`);
        return errors.length === 0
            ? []
            : [
                {
                    title: `The ${shared_1.workspaceFileName()} file is out of sync`,
                    bodyLines: errorGroupBodyLines
                    /**
                     * TODO(JamesHenry): Add support for error documentation
                     */
                    // slug: 'project-has-no-files'
                }
            ];
    }
    filesWithoutProjects() {
        const allFilesFromProjects = this.allProjectFiles();
        const allFilesWithoutProjects = minus(this.files, allFilesFromProjects);
        const first5FilesWithoutProjects = allFilesWithoutProjects.length > 5
            ? allFilesWithoutProjects.slice(0, 5)
            : allFilesWithoutProjects;
        const errorGroupBodyLines = first5FilesWithoutProjects.map(f => `${output_1.output.colors.gray('-')} ${f}`);
        return first5FilesWithoutProjects.length === 0
            ? []
            : [
                {
                    title: `The following file(s) do not belong to any projects:`,
                    bodyLines: errorGroupBodyLines
                    /**
                     * TODO(JamesHenry): Add support for error documentation
                     */
                    // slug: 'file-does-not-belong-to-project'
                }
            ];
    }
    allProjectFiles() {
        return this.projectNodes.reduce((m, c) => [...m, ...c.files], []);
    }
}
exports.WorkspaceIntegrityChecks = WorkspaceIntegrityChecks;
function minus(a, b) {
    return a.filter(aa => b.indexOf(aa) === -1);
}
